xarray/core/dataset.py

```diff
@@ -1095,6 +1095,88 @@ def maybe_chunk(name, var, chunks):
                                  for k, v in self.variables.items()])
         return self._replace_vars_and_dims(variables)
 
+    def _validate_indexers(self, indexers):
+        """ Here we make sure
+        + indexer has a valid keys
+        + indexer is in a valid data type
+        + raise an Error for some confusing case.
+        """
+        from .dataarray import DataArray
+
+        invalid = [k for k in indexers if k not in self.dims]
+        if invalid:
+            raise ValueError("dimensions %r do not exist" % invalid)
+
+        for k, v in iteritems(indexers):
+            dims = getattr(v, 'dims', None)
+            if (dims is not None and k not in dims and
+                    any([d not in indexers for d in dims]) and
+                    any([d in self.dims for d in dims])):
+                raise ValueError(
```

Comment:
Here, I think we should disallow the following indexing,
```python
data = xr.DataArray(np.random.randn(4, 2), dims=['x', 'y'],
                    coords={'x': [0, 1, 2, 3], 'y': ['a', 'b']})
ind = xr.DataArray([1, 0], dims=['y'], coords={'y': [0.0, 1.1]})

data.isel(x=ind)
```
which is grammatically OK but highly error prone.

Reply:
I agree that this case should not be allowed, but I thought we already had logic that covered it, so I'm not sure we need this special case. If I understand correctly, the conflicting labels for `y` should already result in an error when we call `merge_variables`.

(I will take a more careful look at this again later.)

Reply:
Clearer example is

```python
v = xr.Variable(['x', 'y', 'z'], np.arange(60).reshape(3, 4, 5))
ind = xr.Variable(['x'], [0, 1, 2])
v[:, ind.data]  # This gives a [3, 3, 5] array
v[:, ind]  # This gives a [3, 5] array
```
Yes, this is an intended behavior, but I am wondering if this is intuitive.
Can we disallow the broadcast between slice and variable?

Reply:
> Yes, this is an intended behavior, but I am wondering if this is intuitive.
Can we disallow the broadcast between slice and variable?

I agree that this behavior could be surprising, but the same could be said for many things done with vectorized indexing. My preference is to allow it because disallowing it would involve adding one additional special rule, and it's best if we have as few rules as possible. (We do actually use have to use more complex logic in the implementation to make this work, but the complexity is hidden from the user.)

Also, I can think of some use cases for this already. For example, suppose you have a list of maximum positions, and now you want to select the maximum values in each position. This can be done very easily in this way, e.g.,
```
var = xr.Variable(['x', 'y'], np.random.randn(3, 10))
ind = var.argmax('y')  # dimensions (x: 3)
var.isel(y=ind)  # dimensions (x: 3)
```
This is actually easier than doing this with NumPy, where you would need to create a dummy array with `np.arange`, e.g.,
```
arr = np.random.randn(3, 10)
ind = arr.argmax(axis=1)  # shape (3,)
arr[np.arange(arr.shape[0]), ind]  # shape (3,)
```

Reply:
> My preference is to allow it because disallowing it would involve adding one additional special rule, and it's best if we have as few rules as possible. 

I see. I agreed. 
I will remove this block later.

I am still afraid that many existing scripts using previous indexing would behave differently after the release.
I may want to raise a warning if users do some confusing indexings, but it would not be a good idea as it will be very disturbing for advanced users.

Do you have any idea?
or is such backward incompatibility still acceptable as 0.10 is semi-major release?

Reply:
I think it is probably OK to break these cases (which are rare), but I'll bring it up in the 0.10 release issue for the perspective of other developers.